docs/architecture/adr-063-core-module-api.md

```diff
@@ -0,0 +1,546 @@
+amounts of state that can't fit into memory.
+
+Modules which use a single `proto.Message` can use the `GenesisSource.ReadMessage` and `GenesisTarget.ReadMessage`
+methods without needing to deal with raw JSON or codecs.
+
+Modules which use streaming genesis can use the
+`GenesisSource.OpenReader` and `GenesisTarget.OpenWriter` methods. In the case of modules using the ORM, all JSON
+handling is done automatically by the ORM and a single line will be sufficient for registering all of that functionality
+(ex. `ormDb.RegisterGenesis(handler)`).
+
+Low-level `ReadRawJSON` and `WriteRawJSON` are provided in case modules use some other sort of legacy genesis logic
+we haven't anticipated.
+
+#### Begin and End Blockers
+
+The `BeginBlock` and `EndBlock` methods will simply take a `context.Context`, because:
+* most modules don't need Tendermint information other than `BlockInfo` so we can eliminate dependencies on specific
+Tendermint versions
+* for the few modules that need Tendermint block headers and/or return validator updates, specific versions of the
+runtime module will provide specific functionality for interacting with the specific version(s) of Tendermint
+supported
+
+In order for `BeginBlock`, `EndBlock` and `InitGenesis` to send back validator updates and retrieve full Tendermint
+block headers, the runtime module for a specific version of Tendermint could provide services like this:
+```go
+type ValidatorUpdateService interface {
+    SetValidatorUpdates(context.Context, []abci.ValidatorUpdate)
+}
+
+type BeginBlockService interface {
+    GetBeginBlockRequest(context.Context) abci.RequestBeginBlock 
+}
+```
+
+We know these types will change at the Tendermint level and that also a very limited set of modules actually need this
+functionality, so they are intentionally kept out of core to keep core limited to the necessary, minimal set of stable
+APIs.
+
+#### Event Listeners
+
+Handlers allow event listeners for typed events to be registered that will be called deterministically whenever those
+events are called in the state machine. Event listeners are functions that take a context, the protobuf message type of
+the event and optionally return an error. If an event listener returns a non-nil error, this error will fail the process
+that emitted the event and revert any state transitions that this process would have caused.
+
+Event listeners for typed events can be added to the handler struct either by populating the `EventListeners` field
+or using generic helper methods:
+```go
+appmodule.AddEventListener(handler, func (ctx context.Context, e *EventReceive)) { ... })
+```
+
+Event listeners provide a standardized alternative to module hooks, such as `StakingHooks`, even though these hooks
+are still supported and described in [ADR 057: App Wiring](./adr-057-app-wiring-1.md)
+
+Only one event listener per event type can be defined per module. Event listeners will be called in a deterministic
```

Comment:
I believe these events are different than client events, if so we should make sure to identify that. This could get confusing

Reply:
These events are both client and internal events. The legacy "untyped" events are client only.

Reply:
oh didnt know this. Is this mentioned in another ADR?do these events get added to consensus if they are used internally?

Reply:
Yes. This is mentioned first in this ADR I think. Maybe it's too big of a scope and we should save it for a separate ADR